rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,322 @@
+so many implementations and is used in so many contexts, some rules need to be established and all the implementations
+must be updated to behave in a consistent way.
+
+None of the proposed changes to the interface are direct though. They all happen through related types and behaviour.
+
+
+#### types.Watch
+```diff
+type Watch struct {
+	Name string
+	Kinds []WatchKind
+	QueueSize int
+	MetricComponent string
++
++	// AllowRejectedKinds enables partial success mode in which the request won't fail if some of the requested Kinds
++	// are not available for watching.
++	AllowRejectedKinds bool
++
++	// KindsRejectedInFlight contains a list of kinds rejected by upstream implementations in a chain of NewWatcher()
++	// calls. Must be empty if AllowRejectedKinds is not set. A kind cannot be on both Kinds and KindsRejectedInFlight
++	// at the same time.
++	KindsRejectedInFlight []WatchKind
+}
+```
+
+The new field `AllowRejectedKinds` should be added to preserve backwards compatibility when a request for an event stream
+crosses service boundary. If an old downstream service unaware of partial success mode connects to an upgraded auth server,
+it won't set this flag. In that case it should get an error if one of the kinds requested for watching is unavailable.
+
+`KindsRejectedInFlight` should be added to accommodate wrapper implementations of `Events` which perform validations
+and call the inner implementation. If `AllowRejectedKinds` is set and the wrapper determines that some of the requested
+kinds cannot be watched, they will be moved from Kinds to this new field and passed to the inner implementation this way.
+
+#### types.WatchStatus
+
+The `Event` struct has fields for a type and an associated resource. Currently, events with type `OpInit` have
+the associated resource field set to `nil`. We should add a new resource type that can be attached to `OpInit` events
+and carry the list of resource kinds what weren't available for watching:
+
+```protobuf
+message WatchStatusV1 {
+  string Kind = 1;
+  string SubKind = 2;
+  string Version = 3;
+  Metadata Metadata = 4;
+  WatchStatusSpecV1 Spec = 5;
+}
+
+message WatchStatusSpecV1 {
+  repeated WatchKind RejectedKinds = 1;
+}
+```
+
+
+#### Rules for `Events` / `NewWatcher()` implementations 
+
+At the beginning of every `NewWatcher(_ context.Context, w types.Watch)` a few invariants must be established:
+1. `w.Kinds` must not be empty.
+2. `w.KindsRejectedInFlight` must be empty unless `w.AllowRejectedKinds` is true.
+3. A kind+subkind cannot be on both `w.Kinds` and `w.KindsRejectedInFlight` at the same time.
+
+These can be implemented in a `CheckAndSetDefaults()` method on `types.Watch`.
+
+If, during validation, an implementation of `NewWatcher()` determines that a requested resource kind cannot be watched,
+moves it from `w.Kinds` to `w.KindsRejectedInFlight` and one of the above rules stops being satisfied, an error must 
```

Comment:
I like that! The `OpInit` resource could look like this then:

```protobuf
message WatchStatusSpecV1 {
  repeated WatchKind WatchedKinds = 1;
}
```

Then, a consumer would need to handle two variants of `OpInit` event:
 - resource missing: this is a response from an old Auth service
 - resource is present: only specified kinds are considered successful

@fspmarshall what do you think about this?

Reply:
While it would be mildly annoying and not really worth it (especially if we return `WatchedKinds` instead of `RejectedKinds` in `OpInit`), any implementation of `NewWatcher` that wraps and delegates to an upstream `Events` can just return a wrapper for the `types.Watcher` instead, which can alter anything it wants.

Reply:
I updated the doc https://github.com/gravitational/teleport/pull/23494/commits/9261e4fdc19478d6d55c54c95a499ce7fc0934bb

```protobuf
message Watch {
   repeated WatchKind Kinds = 1;
   bool AllowPartialSuccess = 2;
}

message WatchStatusSpecV1 {
  repeated WatchKind Kinds = 1;
}
```

This way we get cleaner API and don't need to update wrappers too much.